theano/gof/wrapper.py

```diff
@@ -0,0 +1,449 @@
+        # Then you can create a Wrap with the wrapper defined above and values for attributes.
+        w = Wrap(wp, attr1=1, key2=2.0, field3=3)
+        print(w.attr1, w.key2, w.field3)
+        d = dict(attr1=1, key2=2, field3=-1)
+        w2 = Wrap(wp, **d)
+        print(w2.attr1, w2.key2, w2.field3)
+
+    """
+
+    def __init__(self, wrapper, **kwargs):
+        if not isinstance(wrapper, Wrapper):
+            raise TypeError('Wrap: 1st constructor argument should be a Wrapper.')
+        for field in wrapper.fields:
+            if field not in kwargs:
+                raise TypeError('Wrap: Wrapper attribute "%s" not in Wrap args.' % field)
+        super(Wrap, self).__init__(**kwargs)
+        self.__dict__.update(wrapper=wrapper)
+
+    def __repr__(self):
+        return 'Wrap(%s)' % ', '.join([('%s:%s' % (k, type(self[k]))) for k in sorted(self.keys())])
+
+    def __getattr__(self, key):
+        if key not in self:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        return self[key]
+
+    def __setattr__(self, key, value):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __delitem__(self, key):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __hash__(self):
+        return hash((type(self), self.wrapper) + tuple(
+            # NB: Wrapped data should have been already filtered.
+            self.wrapper.types[i].make_constant(self[self.wrapper.fields[i]]).signature()
+            for i in range(self.wrapper.length)
+        ))
+
+    def __eq__(self, other):
+        return (type(self) == type(other) and self.wrapper == other.wrapper and all(
+            # NB: Wrapped data should have been already filtered.
+            self.wrapper.types[i].values_eq(self[self.wrapper.fields[i]], other[self.wrapper.fields[i]])
+            for i in range(self.wrapper.length)
+        ))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class Wrapper(Type):
+    """
+    This class can create a struct of Theano types (like TensorType, GpuArrayType, etc.)
+    to be used as a convenience op parameter wrapping many data.
+
+    Wrapper constructor takes key-value args.
+    Key will be the name of the attribute in the struct.
+    Value is the Theano type of this attribute, ie. an instance of (a subclass of) :class:`Type`
+    (eg. ``TensorType('int64', (False,))``).
+
+    In a Python code any attribute named ``key`` will be available via::
+
+        structObject.key
+
+    In a C code, attributes created to represent an instance of the type associated to ``key`` will be available via:
+
+    .. code-block:: c
+
+        structObject->key;
+        structObject->dtype_key; // e.g. from TensorType C code.
+        structObject->other_attribute_named_from_key;
```

Comment:
This seems to suggest that you can have multiple attributes from a single param.  I don't think this is the case.